File: backend/app/services/save_feedback.py
"""
ApeAI — Feedback Storage Service

Handles saving normalized FeedbackItems to Supabase.
Supports both single insert and batch insert (for CSV uploads).
"""

import logging
from typing import Dict, Any, List

from backend.app.db.supabase_client import get_supabase_client
from backend.app.models.feedback import FeedbackItem

logger = logging.getLogger(__name__)


async def save_feedback(item: FeedbackItem) -> Dict[str, Any]:
    """
    Save a single normalized feedback item to Supabase.
    
    Args:
        item: A normalized FeedbackItem ready for storage.
        
    Returns:
        The inserted row data (including generated UUID).
        
    Raises:
        RuntimeError: If Supabase is not configured or insert fails.
    """
    client = get_supabase_client()
    db_data = item.to_db_dict()

    logger.info(
        f"💾 Saving feedback: source={item.source}, "
        f"author={item.author}, "
        f"content_length={len(item.content)} chars"
    )

    try:
        response = client.table("feedback").insert(db_data).execute()
        
        if response.data:
            saved = response.data[0]
            logger.info(f"✅ Feedback saved: id={saved.get('id')}")
            return saved
        else:
            raise RuntimeError("Supabase insert returned no data")

    except Exception as e:
        logger.error(f"❌ Failed to save feedback: {e}")
        raise RuntimeError(f"Failed to save feedback: {e}")


async def save_feedback_batch(items: List[FeedbackItem]) -> List[Dict[str, Any]]:
    """
    Save multiple feedback items in a single batch insert.
    
    Used primarily for CSV uploads to reduce database round trips.
    
    Args:
        items: List of normalized FeedbackItems.
        
    Returns:
        List of inserted rows.
    """
    if not items:
        return []

    client = get_supabase_client()
    db_data = [item.to_db_dict() for item in items]

    logger.info(f"💾 Batch saving {len(items)} feedback items")

    try:
        response = client.table("feedback").insert(db_data).execute()

        if response.data:
            logger.info(f"✅ Batch save complete: {len(response.data)} items saved")
            return response.data
        else:
            raise RuntimeError("Supabase batch insert returned no data")

    except Exception as e:
        logger.error(f"❌ Batch save failed: {e}")
        raise RuntimeError(f"Batch save failed: {e}")


async def get_feedback_stats() -> Dict[str, Any]:
    """
    Get feedback count statistics grouped by source.
    
    Returns:
        Dict with total count and per-source breakdown.
    """
    client = get_supabase_client()

    try:
        response = client.table("feedback").select("source").execute()

        if not response.data:
            return {"total": 0, "by_source": {}}

        # Count per source
        by_source: Dict[str, int] = {}
        for row in response.data:
            src = row.get("source", "unknown")
            by_source[src] = by_source.get(src, 0) + 1

        return {
            "total": len(response.data),
            "by_source": by_source,
        }

    except Exception as e:
        logger.error(f"Failed to get feedback stats: {e}")
        return {"total": 0, "by_source": {}, "error": str(e)}

File: backend-node/src/services/saveFeedback.ts
/**
 * ApeAI — Feedback Storage Service
 *
 * Handles saving normalized FeedbackItems to Supabase.
 * Mirrors Python services/save_feedback.py.
 */

import { getSupabaseClient } from '../db/supabase';
import { FeedbackItem, feedbackItemToDbDict } from '../types/feedback';

export async function saveFeedback(item: FeedbackItem): Promise<Record<string, unknown>> {
  const client = getSupabaseClient();
  const dbData = feedbackItemToDbDict(item);

  console.info(
    `💾 Saving feedback: source=${item.source}, author=${item.author}, content_length=${item.content.length} chars`,
  );

  const { data, error } = await client.from('feedback').insert(dbData).select().single();

  if (error) {
    console.error('❌ Failed to save feedback:', error);
    throw new Error(`Failed to save feedback: ${error.message}`);
  }

  console.info(`✅ Feedback saved: id=${(data as Record<string, unknown>)['id']}`);
  return data as Record<string, unknown>;
}

export async function saveFeedbackBatch(
  items: FeedbackItem[],
): Promise<Array<Record<string, unknown>>> {
  if (items.length === 0) return [];

  const client = getSupabaseClient();
  const dbData = items.map(feedbackItemToDbDict);

  console.info(`💾 Batch saving ${items.length} feedback items`);

  const { data, error } = await client.from('feedback').insert(dbData).select();

  if (error) {
    console.error('❌ Batch save failed:', error);
    throw new Error(`Batch save failed: ${error.message}`);
  }

  console.info(`✅ Batch save complete: ${(data ?? []).length} items saved`);
  return (data ?? []) as Array<Record<string, unknown>>;
}

export async function getFeedbackStats(
  userId?: string,
): Promise<{ total: number; by_source: Record<string, number> }> {
  const client = getSupabaseClient();

  let query = client.from('feedback').select('source');
  if (userId) {
    query = query.eq('user_id', userId);
  }

  const { data, error } = await query;

  if (error) {
    console.error('Failed to get feedback stats:', error);
    return { total: 0, by_source: {} };
  }

  if (!data || data.length === 0) {
    return { total: 0, by_source: {} };
  }

  const bySource: Record<string, number> = {};
  for (const row of data) {
    const src = (row as Record<string, string>)['source'] ?? 'unknown';
    bySource[src] = (bySource[src] ?? 0) + 1;
  }

  return { total: data.length, by_source: bySource };
}

export async function getRecentFeedbacks(
  userId?: string,
  limit = 50,
): Promise<Array<Record<string, unknown>>> {
  const client = getSupabaseClient();

  let query = client
    .from('feedback')
    .select('*')
    .order('created_at', { ascending: false })
    .limit(limit);

  if (userId) {
    query = query.eq('user_id', userId);
  }

  const { data, error } = await query;

  if (error) {
    console.error('Failed to fetch recent feedbacks:', error);
    return [];
  }

  return (data ?? []) as Array<Record<string, unknown>>;
}
